components/layout_2020/display_list/clip_path.rs

```diff
@@ -0,0 +1,228 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cmp::Ordering;
+
+use style::values::computed::basic_shape::{BasicShape, ClipPath};
+use style::values::computed::length::Length;
+use style::values::computed::length_percentage::{LengthPercentage, NonNegativeLengthPercentage};
+use style::values::computed::position::Position;
+use style::values::generics::basic_shape::{GenericShapeRadius, ShapeBox, ShapeGeometryBox};
+use style::values::generics::position::GenericPositionOrAuto;
+use webrender_api::units::{LayoutPoint, LayoutRect, LayoutSize};
+use webrender_api::ClipChainId;
+use webrender_traits::display_list::ScrollTreeNodeId;
+
+use super::{compute_marginbox_radius, normalize_radii};
+
+pub(super) fn build(
+    clip_path: ClipPath,
+    display_list: &mut super::DisplayList,
+    parent_scroll_node_id: ScrollTreeNodeId,
+    parent_clip_chain_id: ClipChainId,
+    fragment_builder: super::BuilderForBoxFragment,
+) -> Option<ClipChainId> {
+    let geometry_box = match clip_path {
+        ClipPath::Shape(_, ShapeGeometryBox::ShapeBox(shape_box)) => shape_box,
+        ClipPath::Shape(_, ShapeGeometryBox::ElementDependent) => ShapeBox::BorderBox,
+        ClipPath::Box(ShapeGeometryBox::ShapeBox(shape_box)) => shape_box,
+        ClipPath::Box(ShapeGeometryBox::ElementDependent) => ShapeBox::BorderBox,
+        _ => return None,
+    };
+    let layout_rect = match geometry_box {
+        ShapeBox::BorderBox => fragment_builder.border_rect,
+        ShapeBox::ContentBox => *fragment_builder.content_rect(),
+        ShapeBox::PaddingBox => *fragment_builder.padding_rect(),
+        ShapeBox::MarginBox => *fragment_builder.margin_rect(),
+    };
+    if let ClipPath::Shape(shape, _) = clip_path {
+        match *shape {
+            BasicShape::Circle(_) | BasicShape::Ellipse(_) | BasicShape::Rect(_) => {
+                build_simple_shape(
+                    *shape,
+                    layout_rect,
+                    parent_scroll_node_id,
+                    parent_clip_chain_id,
+                    display_list,
+                )
+            },
+            BasicShape::Polygon(_) | BasicShape::PathOrShape(_) => None,
+        }
+    } else {
+        create_rect_clip_chain(
+            match geometry_box {
+                ShapeBox::MarginBox => compute_marginbox_radius(
+                    fragment_builder.border_radius,
+                    layout_rect.size(),
+                    fragment_builder.fragment,
+                ),
+                _ => fragment_builder.border_radius,
+            },
+            layout_rect,
+            parent_scroll_node_id,
+            parent_clip_chain_id,
+            display_list,
+        )
+    }
+}
+
+fn build_simple_shape(
+    shape: BasicShape,
+    layout: LayoutRect,
```

Comment:
A layout is more than just a rectangle, so I think it's good to specify to help the reader:

```suggestion
    layout_rect: LayoutRect,
```